contracts/external/callees/Trade0x.sol

```diff
@@ -0,0 +1,113 @@
+/**
+ * SPDX-License-Identifier: UNLICENSED
+ */
+pragma solidity 0.6.10;
+
+pragma experimental ABIEncoderV2;
+
+import {CalleeInterface} from "../../interfaces/CalleeInterface.sol";
+import {IZeroXExchange} from "../../interfaces/ZeroXExchangeInterface.sol";
+import {SafeERC20} from "../../packages/oz/SafeERC20.sol";
+import {ERC20Interface} from "../../interfaces/ERC20Interface.sol";
+import {WETH9} from "../canonical-weth/WETH9.sol";
+
+/**
+ * @author Opyn Team
+ * @title Trade0x
+ * @notice callee contract to trade on 0x.
+ */
+contract Trade0x is CalleeInterface {
+    using SafeERC20 for ERC20Interface;
+    IZeroXExchange public exchange;
+    ERC20Interface public weth;
+    address public assetProxy;
+    address public staking;
+    ///@dev 0x portocal fee to fill 1 order
+    uint256 private PORTOCAL_FEE_BASE = 70000;
+
+    constructor(
+        address _exchange,
+        address _assetProxy,
+        address _weth,
+        address _staking
+    ) public {
+        exchange = IZeroXExchange(_exchange);
+        assetProxy = _assetProxy;
+        weth = ERC20Interface(_weth);
+        weth.safeApprove(_assetProxy, uint256(-1));
+        weth.safeApprove(_staking, uint256(-1));
+    }
+
+    event Trade(
+        address indexed taker,
+        address indexed maker,
+        address takerAsset,
+        address makerAsset,
+        uint256 takerAmount,
+        uint256 makerAmount
+    );
+
+    /**
+     * @dev fill 0x order
+     * @param _sender the original sender who wants to trade on 0x
+     * @param _data abi-encoded order, fillamount, signature and feePayer. fee payer is the address we pull weth from.
+     */
+    function callFunction(address payable _sender, bytes memory _data) external override {
+        (
+            IZeroXExchange.Order memory order,
+            uint256 takerAssetFillAmount,
+            bytes memory signature,
+            address feePayer
+        ) = abi.decode(_data, (IZeroXExchange.Order, uint256, bytes, address));
```

Comment:
This looks good and work! but do u wanna do it in a struct and insure that the struct length is always equal to numbers of args we want ? something similar to this: https://github.com/opynfinance/GammaProtocol/blob/master/contracts/tests/FlashUnwrap.sol#L26
https://github.com/opynfinance/GammaProtocol/blob/master/contracts/tests/FlashUnwrap.sol#L41
https://github.com/opynfinance/GammaProtocol/blob/master/contracts/tests/FlashUnwrap.sol#L43

Reply:
I don't think it's necessary, because if the length is wrong the decode function will revert.

Reply:
yeah, I so that in dYdX implementation and thought it is cool thing, making sure always we are not getting any junk in the data field... but yeah not necessary.